mobilecoind-json/src/data_types.rs

```diff
@@ -491,21 +491,63 @@ pub struct JsonMaskedAmount {
     pub commitment: String,
     pub masked_value: JsonU64,
     pub masked_token_id: String,
+    pub version: Option<u32>,
 }
 
-impl From<&MaskedAmount> for JsonMaskedAmount {
-    fn from(src: &MaskedAmount) -> Self {
-        Self {
-            commitment: hex::encode(src.get_commitment().get_data()),
-            masked_value: JsonU64(src.get_masked_value()),
-            masked_token_id: hex::encode(src.get_masked_token_id()),
+impl From<&mc_api::external::TxOut_oneof_masked_amount> for JsonMaskedAmount {
+    fn from(src: &mc_api::external::TxOut_oneof_masked_amount) -> Self {
+        match src {
+            mc_api::external::TxOut_oneof_masked_amount::v1(src) => Self {
+                commitment: hex::encode(src.get_commitment().get_data()),
+                masked_value: JsonU64(src.get_masked_value()),
+                masked_token_id: hex::encode(src.get_masked_token_id()),
+                version: None,
+            },
+            mc_api::external::TxOut_oneof_masked_amount::v2(src) => Self {
+                commitment: hex::encode(src.get_commitment().get_data()),
+                masked_value: JsonU64(src.get_masked_value()),
+                masked_token_id: hex::encode(src.get_masked_token_id()),
+                version: Some(2),
+            },
+        }
+    }
+}
+
+// Helper conversion between json and protobuf
+impl TryFrom<&JsonMaskedAmount> for mc_api::external::TxOut_oneof_masked_amount {
+    type Error = String;
+
+    fn try_from(
+        src: &JsonMaskedAmount,
+    ) -> Result<mc_api::external::TxOut_oneof_masked_amount, String> {
+        let mut commitment = CompressedRistretto::new();
+        commitment.set_data(
+            hex::decode(&src.commitment)
+                .map_err(|err| format!("Failed to decode commitment hex: {}", err))?,
+        );
+        let mut masked_amount = MaskedAmount::new();
+        masked_amount.set_commitment(commitment);
+        masked_amount.set_masked_value(src.masked_value.into());
+        masked_amount.set_masked_token_id(
+            hex::decode(&src.masked_token_id)
+                .map_err(|err| format!("Failed to decode masked token id hex: {}", err))?,
+        );
+
+        match src.version {
+            None | Some(1) => Ok(mc_api::external::TxOut_oneof_masked_amount::v1(
+                masked_amount,
+            )),
+            Some(2) => Ok(mc_api::external::TxOut_oneof_masked_amount::v2(
+                masked_amount,
+            )),
+            Some(other) => Err(format!("Unknown masked amount version: {}", other)),
         }
     }
 }
 
 #[derive(Deserialize, Serialize, Default, Debug, Clone)]
 pub struct JsonTxOut {
-    pub masked_amount: JsonMaskedAmount,
+    pub masked_amount: Option<JsonMaskedAmount>,
```

Comment:
good point, will circle back on this

Reply:
it's optional because otherwise there is no way to implement `From<external::TxOut>` for this type without panicking

```
impl From<&mc_api::external::TxOut> for JsonTxOut {
    fn from(src: &mc_api::external::TxOut) -> Self {
        Self {
            masked_amount: src.masked_amount.as_ref().map(Into::into),
            target_key: hex::encode(src.get_target_key().get_data()),
            public_key: hex::encode(src.get_public_key().get_data()),
            e_fog_hint: hex::encode(src.get_e_fog_hint().get_data()),
            e_memo: hex::encode(src.get_e_memo().get_data()),
        }
    }
}
```